### DISCUSSION TITLE
why you select wayland ？The root cause, I  look forward to your detailed reply. Thanks very much.

### DISCUSSION THREAD

**Discussion by 9Wr** [ravynsoft/ravynos#380] [General]
hello，I want to know  you select wayland 、wlroots、OpenGL ES2、EGL DRM KMS**  replace  Xorg ，why ？？
Maybe
1、 keep compatible with the mac os applications(Appkit)? 
2、 some drawbacks on X11?
3、For UX？
........
I have no idea from your description：
“Our previous images used a pretty traditional X11/KDE desktop environment that was shoehorned into looking and behaving like the Mac desktop.
However, this can only go so far due to the very different architecture and design of the two systems. ”

The root cause, I  look forward to your detailed reply. Thanks very much.

**Comment by Anvith00** [comment]
If you would not mind, I would recommend joining the discord server/matrix rooms since the contents of the webpage are not quite up to date as much as the server/ rooms are, there's been some recent discussions on the display server stuff in the ravynOS discord server/ or its Matrix rooms (each bridged to the channel of the ravynOS discord server). Also that the developers are more likely to respond in discord/ matrix.

For example; there's been a recent decision over making use of Xorg and running the Wayland window server inside it (for virtual machines), because there have been some hiccups with using Wayland to accelerate graphics on virtual machines.

**Reply by 9Wr** [reply]
Thank you for your reply.
I am very interested in RavynOS and hope to join some technical discussion groups.
Look forward to your recommendation

**Reply by mszoek** [reply]
OK, I'll try to explain briefly why I switched away from X11 to a different stack.
X11 does not really know about applications - only the windows they create. And it doesn't know which application owns a window except through a voluntary and insecure mechanism. The window manager is responsible for decorating, moving, minimizing etc the windows but everything is optional in X. It's all just "hints" that can be ignored.

WindowServer needs to reliably know which process/app owns which windows. It needs to be able to send messages to them. It needs to tell Dock about apps that start and quit, and the windows they create and destroy, miniaturize and de-miniaturize, etc. It needs to track which app is active in order to display the right global menus and other context-sensitive items. This was all sort of possible with X but it needed really wonky tricks and wasn't very solid. By starting over and writing the whole display server, it will behave the Mac way. I want this for the best chance at actual compatibility.

I chose to use wayland both because it has some existing compositor functions and because it's a standard we can use to support FreeBSD/Linux apps either natively on wayland or with X-on-wayland. It's a pain in the butt, and I hope it will be worth it.

**Comment by 9Wr** [comment]
ravynOS, Always
We're building familiar apps native to ravynOS, packaged as portable .app bundles and built with our core technologies of Cocoa, Qt, and Mach. These are guaranteed to work and fit seamlessly into your desktop.

“core technologies of Cocoa, Qt, and Mach.”   ： What are the core technologies? what are the difficulties？

**Reply by mszoek** [reply]
The core stack is Cocoa (Objective C), Core frameworks (C) and Mach (C). Difficulties include:
* that we still use ELF, not Mach-O so tools have bad expectations.
* there's no XCode on ravynOS yet so you may need a Mac for UI design
* Cocoa and Core are very incomplete, lots of stub functions
* the look of Cocoa right now is ugly
